### PR TITLE
[core][observability] Refactor ray log API

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This PR changes a few `ray log` semantics based on this: https://docs.google.com/document/d/1mwLz589IZ4LlPh218dDTMskec9hp3r40hw7ROYq3eVo/edit
-  Change `ray logs` with various ids to `ray logs <subcommand>` 
   - i.e. `ray logs worker --pid=x` and `ray logs actor --id=x` and `ray logs cluster <file_name>` 
- Added suffix options for querying logs through pid/actor id to differentiate `.out` and `.err` files. 
- Alias `ray logs ...` to `ray logs cluster ...` so that 
  - `ray logs` : print help 
  - `ray logs cluster`: show all logs on head node 
  - `ray logs <glob>` same as `ray logs cluster <glob>`: list/get files by filename. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/state/state_cli.py [hunk 1]
```diff
@@ -676,3 +680,426 @@ def object_summary(ctx, timeout: float, address: str):
...
+        ```
+        ray logs actor --id ABC --follow
+        ```
+
+        Get the actor log from pid 123, ip ABC.
+        Note that this goes well with the driver log of Ray which prints
+        (ip=ABC, pid=123, class_name) logs.
+
+        ```
+        ray logs actor --pid=123  —ip=ABC
```
Comment (rkooo567) [inline python/ray/experimental/state/state_cli.py:1019]: Add an suffix example?

File: python/ray/experimental/state/state_cli.py [hunk 2]
```diff
@@ -676,3 +680,426 @@ def object_summary(ctx, timeout: float, address: str):
...
+    suffix: str,
+):
+    """Get/List logs associated with a worker process.
+
+    Example:
+
+        Follow the log file from a worker process with pid=ABC.
+
+        ```
+        ray logs worker --pid ABC --follow
```
Comment (rkooo567) [inline python/ray/experimental/state/state_cli.py:1086]: same for suffix example

File: python/ray/experimental/state/state_cli.py [hunk 3]
```diff
@@ -676,3 +680,426 @@ def object_summary(ctx, timeout: float, address: str):
...
+    default=None,
+    help="Filters the logs by this NodeID",
+)
+
+log_suffix_option = click.option(
+    "--suffix",
+    required=False,
+    default="out",
+    type=click.Choice(["out", "err"], case_sensitive=False),
+    help="The suffix of the log file that denotes the log type.",
```
Comment (rkooo567) [inline python/ray/experimental/state/state_cli.py:744]: Can you mention out means logs to stdout, and err means logs to stderr?

**Review by rkooo567 — APPROVED** [review]
Some nits

**Comment by rkooo567** [comment]
Lint failure

**Comment by rkooo567** [comment]
also can you get the doc approval?

File: python/ray/experimental/state/api.py [hunk 4]
```diff
@@ -1140,11 +1141,13 @@ def get_log(
         filename: Name of the file (relative to the ray log directory) to be retrieved.
         actor_id: Id of the actor if getting logs from an actor.
         task_id: Id of the task if getting logs generated by a task.
-        pid: PID of the worker if getting logs generated by a worker.
+        pid: PID of the worker if getting logs generated by a worker. When querying
+            with pid, one of node_id or node_ip has to supplied.
```
Comment (dmatrix) [inline python/ray/experimental/state/api.py:1145]: May be say, 
When querying with pid, either node_id or node_ip must be supplied.

File: python/ray/experimental/state/common.py [hunk 5]
```diff
@@ -292,8 +294,8 @@ def __post_init__(self):
             raise ValueError(f"Invalid media type: {self.media_type}")
         if not (self.node_id or self.node_ip) and not (self.actor_id or self.task_id):
             raise ValueError(
-                "node_id or node_ip should be provided."
-                "Please provide at least one of them."
+                "node_id or node_ip should be provided when no actor or task id "
+                "is provided. Please provide at least one of them."
```
Comment (dmatrix) [inline python/ray/experimental/state/common.py:298]: provided is used twice. Perhaps use another word:
"node_id or node_ip must be provided as constructor arguments when no actor or task_id is supplied as arguments."

**Review by dmatrix — APPROVED** [review]
made minor nit or suggestions. Otherwise LGTM

**Comment by rkooo567** [comment]
The tests already passed, and new changes are just doc changes.

File: doc/source/ray-observability/state/state-api.rst [hunk 6]
```diff
@@ -422,14 +422,14 @@ Logs
...
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+E.g., Get all retrievable log file names from a head node in a cluster
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 .. tabbed:: CLI
 
     .. code-block:: bash
 
-        ray logs 
+        ray logs cluster
```
Comment (rkooo567) [inline doc/source/ray-observability/state/state-api.rst:432]: Btw it is confusing from the doc that only the first example is using `ray logs cluster` and other examples start using `ray logs <file>`. We should fix this as a follow up (use a consistent example and put a note ray logs is a alias of ray logs cluster)

**Pull request merged**